config.org
* Better Emacs
** bind-file
    #+BEGIN_SRC emacs-lisp
      (defmacro bind-file (binding file)
	`(global-set-key (kbd ,binding) (lambda () (interactive) (find-file ,file))))
    #+END_SRC
** COMMENT Hydra
    #+BEGIN_SRC emacs-lisp
      (use-package hydra
  :straight t)
    #+END_SRC
** Helpful
    #+BEGIN_SRC emacs-lisp
      (use-package helpful
	:straight (helpful :type git :host github :repo "Wilfred/helpful")
	:config 
	(global-set-key (kbd "C-h f") #'helpful-callable)
	(global-set-key (kbd "C-h v") #'helpful-variable)
	(global-set-key (kbd "C-h k") #'helpful-key))
    #+END_SRC
** Ibuffer groups
    #+BEGIN_SRC emacs-lisp
      (setq ibuffer-saved-filter-groups
		(quote (("default"
			 ("dired" (mode . dired-mode))
			 ("perl" (mode . cperl-mode))
			 ("erc" (mode . erc-mode))
			 ("python" (mode . python-mode))
			 ("planner" (or
				     (name . "^\\*Calendar\\*$")
				     (name . "^diary$")
				     (mode . muse-mode)))
			 ("emacs" (or
				   (name . "^\\*scratch\\*$")
				   (name . "^\\*Messages\\*$")))
			 ("gnus" (or
				  (mode . message-mode)
				  (mode . bbdb-mode)
				  (mode . mail-mode)
				  (mode . gnus-group-mode)
				  (mode . gnus-summary-mode)
				  (mode . gnus-article-mode)
				  (name . "^\\.bbdb$")
				  (name . "^\\.newsrc-dribble")))))))
    #+END_SRC
** Profiling
    #+BEGIN_SRC emacs-lisp
      (use-package esup
	:straight t)
    #+END_SRC
** no-littering
    #+BEGIN_SRC emacs-lisp
      (use-package no-littering
	:straight t)
    #+END_SRC
* Editing
** EVIL
   #+BEGIN_SRC emacs-lisp
     (use-package evil
       :straight t
       :config
       (evil-mode 1))
   #+END_SRC
*** evil-magit
   #+BEGIN_SRC emacs-lisp
     (use-package evil-magit
       :straight t)
       #+END_SRC
*** Keyboard bindings
#+BEGIN_SRC emacs-lisp
  (use-package evil-leader
    :straight t
    :config
    (global-evil-leader-mode t)
    (evil-leader/set-leader "<SPC>")
    (evil-leader/set-key
      "s s" 'swiper
      "f f" 'counsel-find-file
      "f s" 'save-buffer
      "g s" 'magit
      "f U" 'undo-tree-visualize
      "b b" 'counsel-switch-buffer
      "b i" 'ibuffer
      "<SPC>" 'counsel-M-x
      "'" 'eshell
      "c c" 'compile
      ;; wm
      "w s" 'split-window-below
      "w v" 'split-window-horizontally
      "q q" 'save-buffers-kill-emacs
      ))
#+END_SRC
*** evil collection
#+BEGIN_SRC emacs-lisp
     (use-package evil-collection
       :straight t
       :custom (evil-collection-setup-minibuffer t)
       :init (evil-collection-init))
#+END_SRC
*** org
#+BEGIN_SRC emacs-lisp
     (use-package evil-org
       :straight t
       :config
       (evil-org-set-key-theme
	'(textobjects insert navigation additional shift todo heading))
       (add-hook 'org-mode-hook (lambda () (evil-org-mode))))
#+END_SRC
*** else
#+BEGIN_SRC emacs-lisp
     (use-package evil-surround
       :straight t
       :config (global-evil-surround-mode))

     (use-package evil-indent-textobject
       :straight t)
#+END_SRC
** History
#+BEGIN_SRC emacs-lisp
     (use-package undo-tree
       :straight t
       :config
       (global-undo-tree-mode 1))
#+END_SRC
** Completion
*** IVY/counsel/swiper
**** IVY
#+BEGIN_SRC emacs-lisp
      (use-package ivy
	:straight t
	:config
	(ivy-mode 1))
#+END_SRC
***** IVY rich
#+BEGIN_SRC emacs-lisp
      (use-package ivy-rich
	:straight t
	:after counsel
	:init
	(ivy-rich-mode 1))
#+END_SRC
**** Counsel
#+BEGIN_SRC emacs-lisp
      (use-package counsel
	:straight t)
#+END_SRC
***** counsel tramp
#+BEGIN_SRC emacs-lisp
      (use-package counsel-tramp
	:straight t)
#+END_SRC
*** Snippets
#+BEGIN_SRC emacs-lisp
      (use-package yasnippet
	:diminish yas-minor-mode
	:init
	(use-package yasnippet-snippets :after yasnippet)
	:hook ((prog-mode LaTeX-mode org-mode) . yas-minor-mode)
	:bind
	(:map yas-minor-mode-map ("C-c C-n" . yas-expand-from-trigger-key))
	(:map yas-keymap
	      (("TAB" . smarter-yas-expand-next-field)
	       ([(tab)] . smarter-yas-expand-next-field)))
	:config
	(yas-reload-all)
	(defun smarter-yas-expand-next-field ()
	  "Try to `yas-expand' then `yas-next-field' at current cursor position."
	  (interactive)
	  (let ((old-point (point))
		(old-tick (buffer-chars-modified-tick)))
	    (yas-expand)
	    (when (and (eq old-point (point))
		       (eq old-tick (buffer-chars-modified-tick)))
	      (ignore-errors (yas-next-field))))))
#+END_SRC
*** Company
#+BEGIN_SRC emacs-lisp
      (use-package company
	:straight t
	:custom
	(company-show-numbers t)
	(company-minimum-prefix-length 1)
	(company-idle-delay 0)
	:config
	(global-company-mode 1))
#+END_SRC
**** Company-lsp 
#+BEGIN_SRC emacs-lisp
      (use-package company-lsp
	:straight t
	:config
	(push 'company-lsp company-backends))
#+END_SRC
*** TODO Abbrev
*** TODO dabbrev
** Navigation
*** Jump
#+BEGIN_SRC emacs-lisp
     (use-package avy
       :straight t
       :custom
       (evil-leader/set-key 
	 "j w" 'avy-goto-word-1
	 "j l" 'avy-goto-line))
#+END_SRC
*** TODO Outline
*** Window management
**** Winum
#+BEGIN_SRC emacs-lisp
     (use-package winum
       :straight t
       :config
       (winum-mode 1)
       (evil-leader/set-key 
	 "0" 'winum-select-window-0-or-10
	 "1" 'winum-select-window-1
	 "2" 'winum-select-window-2
	 "3" 'winum-select-window-3
	 "4" 'winum-select-window-4
	 "5" 'winum-select-window-5
	 "6" 'winum-select-window-6
	 "7" 'winum-select-window-7
	 "8" 'winum-select-window-8))
#+END_SRC
*** Search
**** swiper
#+BEGIN_SRC emacs-lisp
  (use-package swiper
    :straight t
    :bind (("C-s" . swiper)))
#+END_SRC

**** ripgrep
** Writing
- [[https://explog.in/notes/writingsetup.html][write-up]]
*** Auto-fill
#+BEGIN_SRC emacs-lisp
      (add-hook
       'text-mode-hook
       'auto-fill-mode)
#+END_SRC
*** Center text
#+BEGIN_SRC emacs-lisp
      (add-hook
       'text-mode-hook
       'olivetti-mode)
#+END_SRC
** File short cuts
#+BEGIN_SRC emacs-lisp
      (global-set-key (kbd "C-x C-.") (lambda () (interactive) (find-file "~/.emacs.d/config.org")))
#+END_SRC
* App
** Programming
*** Languages
**** Python
#+BEGIN_SRC emacs-lisp
     (use-package python-mode
     :straight t
  :after flycheck
  :mode "\\.py\\'"
  :custom
  (python-indent-offset 4)
  (flycheck-python-pycompile-executable "python3")
  (python-shell-interpreter "python3"))
#+END_SRC
***** elpy
- [[https://elpy.readthedocs.io/en/latest/ide.html][documentation]]
  #+BEGIN_SRC emacs-lisp
       (use-package elpy
	 :straight t
	 :init
	 (elpy-enable))
  #+END_SRC
**** octave
#+BEGIN_SRC emacs-lisp
       (use-package ac-octave
	 :straight t
	 :config
	 (add-to-list 'auto-mode-alist '("\\.m\\'" . octave-maybe-mode)))
#+END_SRC
**** C-C++
#+BEGIN_SRC emacs-lisp
  (add-hook 'c-mode-hook 'lsp)
  (add-hook 'c++-mode-hook 'lsp)
  (setq lsp-clangd-binary-path "~/.guix-profile/bin/clangd")
#+END_SRC
**** Promela
#+BEGIN_SRC emacs-lisp
  (use-package promela-mode
    :straight (promela-mode :type git :host github :repo "rudi/promela-mode")
    ;;:after flycheck
    ;;:hook (company-mode . flycheck-mode)
    :config
    (flycheck-define-checker promela
      "A Promela syntax checker using the Promela compiler"
      :command ("spin" "-C" source)
      :error-patterns
      ((error line-start "spin: " (file-name) ":" line ", Error: " (message) line-end))
      :modes promela-mode)
    ;; (flycheck-mode)
    ;; (flycheck-select-checker 'promela)
    )
#+END_SRC
*** Utils
**** Linting
#+BEGIN_SRC emacs-lisp
       (use-package flycheck
       :straight t
	 :defer t
	 :diminish
	 :hook ((prog-mode markdown-mode) . flycheck-mode)
	 :custom
	 (flycheck-global-modes
	  '(not text-mode outline-mode fundamental-mode org-mode
		diff-mode shell-mode eshell-mode term-mode))
	 (flycheck-emacs-lisp-load-path 'inherit)
	 (flycheck-indication-mode 'right-fringe)
	 :init
	 (use-package flycheck-grammarly :defer t)
  :straight t
	 :config
	 (when (fboundp 'define-fringe-bitmap)
	   (define-fringe-bitmap 'flycheck-fringe-bitmap-double-arrow
	     [16 48 112 240 112 48 16] nil nil 'center))
	 (flycheck-add-mode 'javascript-eslint 'js-mode)
	 (flycheck-add-mode 'typescript-tslint 'rjsx-mode))
#+END_SRC
**** imenu-list
#+BEGIN_SRC emacs-lisp
       (use-package imenu-list
	 :straight t)
#+END_SRC
:config
(evil-leader/set-key
"f m" 'imenu-list))
**** rg
#+BEGIN_SRC emacs-lisp
       (use-package rg
	 :straight t)
#+END_SRC
**** Project management
***** Projectile
#+BEGIN_SRC emacs-lisp
       (use-package projectile
	 :straight t
	 :config
	 (projectile-mode +1)
	 (add-to-list 'projectile-globally-ignored-directories "__pycache__"))
#+END_SRC
****** Counsel-projectile
#+BEGIN_SRC emacs-lisp
       (use-package counsel-projectile
	 :straight t
	 :config
	 (evil-leader/set-key
	   "p p" 'counsel-projectile-switch-project
	   "p f" 'counsel-projectile-find-file))
#+END_SRC
***** prodigy
#+BEGIN_SRC emacs-lisp
       (use-package prodigy
	 :ensure t)
#+END_SRC
*** Programs
**** TODO RestClient
*** TODO LSP
#+BEGIN_SRC emacs-lisp
       (use-package lsp-mode
	 :straight t
	 :defer t
	 :commands lsp
	 :custom
	 (lsp-auto-guess-root nil)
	 (lsp-prefer-flymake nil) ; Use flycheck instead of flymake
	 (lsp-file-watch-threshold 2000)
	 (read-process-output-max (* 1024 1024))
	 (lsp-eldoc-hook nil)
	 :hook ((java-mode python-mode go-mode
			   js-mode js2-mode typescript-mode web-mode
			   c-mode c++-mode objc-mode) . lsp))
#+END_SRC
**** LSP UI
#+BEGIN_SRC emacs-lisp
       (use-package lsp-ui
	 :straight t
	 :after lsp-mode
	 :diminish
	 :commands lsp-ui-mode
	 :custom-face
	 (lsp-ui-doc-background ((t (:background nil))))
	 (lsp-ui-doc-header ((t (:inherit (font-lock-string-face italic)))))
	 :custom
	 (lsp-ui-doc-header t)
	 (lsp-ui-doc-include-signature t)
	 (lsp-ui-doc-border (face-foreground 'default))
	 (lsp-ui-sideline-enable nil)
	 (lsp-ui-sideline-ignore-duplicate t)
	 (lsp-ui-sideline-show-code-actions nil)
	 :config
	 ;; Use lsp-ui-doc-webkit only in GUI
	 (if (display-graphic-p)
	     (setq lsp-ui-doc-use-webkit t))
	 ;; WORKAROUND Hide mode-line of the lsp-ui-imenu buffer
	 ;; https://github.com/emacs-lsp/lsp-ui/issues/243
	 (defadvice lsp-ui-imenu (after hide-lsp-ui-imenu-mode-line activate)
	   (setq mode-line-format nil)))
#+END_SRC
**** DAP
#+BEGIN_SRC emacs-lisp
       (use-package dap-mode
	 :straight t
	 :diminish
	 :bind
	 (:map dap-mode-map
	       (("<f12>" . dap-debug)
		("<f8>" . dap-continue)
		("<f9>" . dap-next)
		("<M-f11>" . dap-step-in)
		("C-M-<f11>" . dap-step-out)
		("<f7>" . dap-breakpoint-toggle))))
#+END_SRC
*** TODO Eshell
- [[http://www.modernemacs.com/post/custom-eshell/][inspo]]
#+BEGIN_SRC emacs-lisp
  (require 'dash)
  (require 's)

  (defmacro with-face (STR &rest PROPS)
    "Return STR propertized with PROPS."
    `(propertize ,STR 'face (list ,@PROPS)))

  (defmacro esh-section (NAME ICON FORM &rest PROPS)
    "Build eshell section NAME with ICON prepended to evaled FORM with PROPS."
    `(setq ,NAME
	   (lambda () (when ,FORM
			(-> ,ICON
			    (concat esh-section-delim ,FORM)
			    (with-face ,@PROPS))))))

  (defun esh-acc (acc x)
    "Accumulator for evaluating and concatenating esh-sections."
    (--if-let (funcall x)
	(if (s-blank? acc)
	    it
	  (concat acc esh-sep it))
      acc))

  (defun esh-prompt-func ()
    "Build `eshell-prompt-function'"
    (concat esh-header
	    (-reduce-from 'esh-acc "" eshell-funcs)
	    "\n"
	    eshell-prompt-string))



  ;; Separator between esh-sections
  (setq esh-sep " | ")  ; or " | "

  ;; Separator between an esh-section icon and form
  (setq esh-section-delim " ")

  ;; Eshell prompt header
  (setq esh-header "\n┌─")  ; or "\n┌─"

  ;; Eshell prompt regexp and string. Unless you are varying the prompt by eg.
  ;; your login, these can be the same.
  (setq eshell-prompt-regexp "└─> ")   ; or "└─> "
  (setq eshell-prompt-string "└─> ")   ; or "└─> "

  (esh-section esh-dir
	       ""  ;  (faicon folder)
	       (abbreviate-file-name (eshell/pwd))
	       '(:foreground "gold" :bold ultra-bold :underline t))

  (esh-section esh-git
	       "";  (git icon)
	       (magit-get-current-branch)
	       '(:foreground "pink"))

  (esh-section esh-clock
	       "";  (clock icon)
	       (format-time-string "%H:%M" (current-time))
	       '(:foreground "forest green"))

  ;; Choose which eshell-funcs to enable
  (setq eshell-funcs (list esh-dir esh-git esh-clock))

  ;; Enable the new eshell prompt
  (setq eshell-prompt-function 'esh-prompt-func)
#+END_SRC
*** GIT
**** Magit
#+BEGIN_SRC emacs-lisp
      (use-package magit
	:straight t
	:bind (("C-x g" . magit)))
#+END_SRC
**** vc-msg
effective git blame
#+BEGIN_SRC emacs-lisp
      (use-package vc-msg
	:straight t
	:config
	(evil-leader/set-key
	  "g b" 'vc-msg-show))
#+END_SRC
** Org-Mode
*** Indentation
#+BEGIN_SRC emacs-lisp
      (setq org-indent-indentation-per-level 1)
      (setq org-adapt-indentation nil)
      (setq org-hide-leading-stars 't)
#+END_SRC
*** Hide details
#+BEGIN_SRC emacs-lisp
    (setq org-hide-emphasis-markers t)
#+END_SRC
*** Details
#+BEGIN_SRC emacs-lisp
      (customize-set-variable 'org-blank-before-new-entry 
			      '((heading . nil)
				(plain-list-item . nil)))
      (setq org-cycle-separator-lines 1)
#+END_SRC
*** Org-main
#+BEGIN_SRC emacs-lisp
      (setq org-main-dir "~/org-mode/"
	    org-main-file (concat org-main-dir "main.org"))
#+END_SRC
**** bind file
#+BEGIN_SRC emacs-lisp
      (bind-file "C-c o" org-main-file)
#+END_SRC
*** Agenda
#+BEGIN_SRC emacs-lisp
(setq org-angenda-files (list org-main-file))
#+END_SRC
(use-package elegant-agenda-mode
:straight (elegant-agenda-mode :type git :host github :repo "justincbarclay/elegant-agenda-mode")
:hook org-agenda-mode-hook)
*** TODOs
#+BEGIN_SRC emacs-lisp
(setq org-todo-keywords
      '((sequence "TODO" "DOING" "REVIEW" "|" "DONE" "CANCELED")
        (sequence "BUG" "DEBUG" "|" "SOLVED")))
#+END_SRC
** Reader
*** elfeed
#+BEGIN_SRC emacs-lisp
  (use-package elfeed
    :defer t) 

  (use-package elfeed-org
    :straight t
    :config
    (defvar elfeed-feeds-alist
      (list (concat org-main-dir "elfeedMain.org"))))

  (use-package elfeed-web
    :straight t)
#+END_SRC
*** eww
**** open-link-with-eww
#+BEGIN_SRC emacs-lisp
      (setq browse-url-browser-function
	    '(("wikipedia\\.org" . eww-browse-url)
	      ("github" . eww-browse-url) ;; TODO remoe the clutter
	      ("." . browse-url-default-browser)))
#+END_SRC
**** TODO improve github go directly to the README.md
- CURRENT_URL + /blob/master/README.md
  - User raw url
    #+BEGIN_SRC emacs-lisp
      (defun eww-goto-readme ()
	(eww-browse-url (concat (plist-get eww-data :url) "/blob/master/README.md")))
    #+END_SRC
**** Preview
***** Org-mode
#+BEGIN_SRC emacs-lisp
      (use-package org-preview-html
	:straight t)
#+END_SRC
***** Markdown
#+BEGIN_SRC emacs-lisp
      (use-package markdown-preview-eww
	:straight t)
#+END_SRC
** Communications
*** ERC
*** GNUS
** PDF
#+BEGIN_SRC emacs-lisp
  (use-package pdf-tools
    :straight t
    :defer t) 
#+END_SRC
* System Management
** GUIX
#+BEGIN_SRC emacs-lisp
  (use-package geiser-guile
	 :straight t)
    (use-package guix
	 :straight t
  :after geiser-guile)
#+END_SRC
** Docker
#+BEGIN_SRC emacs-lisp
  (use-package docker
    :straight t)
  (use-package dockerfile-mode
    :straight t)
  (use-package docker-tramp
    :straight t)
#+END_SRC
* AESTHETIC
** Sanity
Remove annoying things

#+BEGIN_SRC emacs-lisp 
  (set-default 'cursor-type  '(bar . 1))
  (blink-cursor-mode 0)
  ;; remove sound
  (setq visible-bell t)
  (setq ring-bell-function 'ignore)
  ;; yes/no to y/n
  (defalias 'yes-or-no-p 'y-or-n-p)
#+END_SRC
** Emacs look
*** Theme
- add doom solarized
- acario is also good
#+BEGIN_SRC emacs-lisp 
  (use-package doom-themes
    :straight t
    :config
    ;; Global settings (defaults)
    (setq doom-themes-enable-bold t    ; if nil, bold is universally disabled
	  doom-themes-enable-italic t) ; if nil, italics is universally disabled
    (load-theme 'doom-solarized-dark t) 
    ;; or for tremacs users
    ;; (setq doom-themes-treemacs-theme "doom-colors") ; use the colorful treemacs theme
    ;; (doom-themes-treemacs-config)
    ;; Corrects (and improves) org-mode's native fontification.
    (doom-themes-org-config))
#+END_SRC
*** tabs
#+BEGIN_SRC emacs-lisp
  (use-package centaur-tabs
    :straight t
    :demand
    :config
    (centaur-tabs-mode t)
    (setq centaur-tabs-set-bar 'over
	  centaur-tabs-style "wave"
	  centaur-tabs-set-modified-marker t
	  centaur-tabs-modified-marker "*")
    :bind
    (:map evil-normal-state-map
	       ("g t" . centaur-tabs-forward)
	       ("g T" . centaur-tabs-backward)))
#+END_SRC
*** Modeline
#+BEGIN_SRC emacs-lisp
  (set-fontset-font "fontset-default"  '(#x2600 . #x26ff) "Fira Code 16")

  (define-key mode-line-major-mode-keymap [header-line]
    (lookup-key mode-line-major-mode-keymap [mode-line]))

  (defun mode-line-render (left right)
    "Function to render the modeline LEFT to RIGHT."
    (let* ((available-width (- (window-width) (length left) )))
      (format (format "%%s %%%ds" available-width) left right)))

  (setq-default mode-line-format
		'((:eval
		   (mode-line-render
		    (format-mode-line (list
				       (format " %d " (winum-get-number)) ;; winum display
				       (propertize "☰" 'face `(:inherit mode-line-buffer-id)
						   'help-echo "Mode(s) menu"
						   'mouse-face 'mode-line-highlight
						   'local-map   mode-line-major-mode-keymap)
				       " %b "
				       (if (and buffer-file-name (buffer-modified-p))
					   (propertize "(modified)" 'face `(:inherit face-faded)))))
		    (format-mode-line
		     (propertize "%4l:%2c" 'face `(:inherit face-faded)))))))
#+END_SRC
*** Line numbers
   #+BEGIN_SRC emacs-lisp
     (setq display-line-numbers-type 'relative)
     (add-hook 'prog-mode-hook 'display-line-numbers-mode)
   #+END_SRC
*** Splash screen
   #+BEGIN_SRC emacs-lisp
     (setq fancy-splash-image (expand-file-name  "~/.emacs.d/assets/common-lisp.png"))
     (setq inhibit-startup-screen t)
     (setq inhibit-startup-echo-area-message t)
     (setq inhibit-startup-message t)
     (setq inhibit-startup-echo-area-message t)
     (setq warning-minimum-level :emergency)
   #+END_SRC
**** Dashboard
   #+BEGIN_SRC emacs-lisp
     (use-package dashboard
       :straight t
       :config
       (dashboard-setup-startup-hook)
       (setq dashboard-center-content t
	     dashboard-banner-logo-title "Simplex Sigillum Veri"
	     dashboard-items '((recents  . 5))
	     dashboard-footer-messages nil)
       (setq dashboard-startup-banner
	     (expand-file-name
	      (if (not window-system) "welcome.txt"  "common-lisp.png")
	      (expand-file-name "assets" user-emacs-directory))))
   #+END_SRC
** Text
*** Font
- for dyslexia
#+BEGIN_SRC emacs-lisp 
  (set-face-font 'default "Roboto Mono Light 11")
#+END_SRC
*** Highligth
#+BEGIN_SRC emacs-lisp
  (show-paren-mode t)
#+END_SRC
*** icons
#+BEGIN_SRC emacs-lisp
  (use-package all-the-icons
    :straight t)
#+END_SRC


config.el
(use-package flycheck
  :defer t
  :diminish
  :hook ((prog-mode markdown-mode) . flycheck-mode)
  :custom
  (flycheck-global-modes
   '(not text-mode outline-mode fundamental-mode org-mode
	 diff-mode shell-mode eshell-mode term-mode))
  (flycheck-emacs-lisp-load-path 'inherit)
  (flycheck-indication-mode 'right-fringe)
  :init
  (use-package flycheck-grammarly :defer t)
  :config
  (when (fboundp 'define-fringe-bitmap)
    (define-fringe-bitmap 'flycheck-fringe-bitmap-double-arrow
      [16 48 112 240 112 48 16] nil nil 'center))
  (flycheck-add-mode 'javascript-eslint 'js-mode)
  (flycheck-add-mode 'typescript-tslint 'rjsx-mode))

(setq org-indent-indentation-per-level 1)
(setq org-adapt-indentation nil)
(setq org-hide-leading-stars 't)

(setq org-hide-emphasis-markers t)

(customize-set-variable 'org-blank-before-new-entry 
                        '((heading . nil)
                          (plain-list-item . nil)))
(setq org-cycle-separator-lines 1)

(use-package evil
  :config
  (evil-mode 1))

(use-package undo-tree
  :config
  (global-undo-tree-mode 1))

(use-package ivy
  :config
  (ivy-mode 1))

(use-package yasnippet
:diminish yas-minor-mode
:init
(use-package yasnippet-snippets :after yasnippet)
:hook ((prog-mode LaTeX-mode org-mode) . yas-minor-mode)
:bind
(:map yas-minor-mode-map ("C-c C-n" . yas-expand-from-trigger-key))
(:map yas-keymap
      (("TAB" . smarter-yas-expand-next-field)
       ([(tab)] . smarter-yas-expand-next-field)))
:config
(yas-reload-all)
(defun smarter-yas-expand-next-field ()
  "Try to `yas-expand' then `yas-next-field' at current cursor position."
  (interactive)
  (let ((old-point (point))
        (old-tick (buffer-chars-modified-tick)))
    (yas-expand)
    (when (and (eq old-point (point))
               (eq old-tick (buffer-chars-modified-tick)))
      (ignore-errors (yas-next-field))))))

(use-package company
  :custom
  (company-show-numbers t)
  (company-idle-delay 0)
  :config
  (global-company-mode 1))

(use-package avy
    :custom
    (global-set-key (kbd "C-:") 'avy-goto-word-1))

(use-package swiper
  :bind (("C-f" . swiper)))

(add-hook
 'text-mode-hook
 'auto-fill-mode)

(add-hook
 'text-mode-hook
 'olivetti-mode)

(use-package doom-modeline
  :ensure t
  :init (doom-modeline-mode 1))

(use-package dashboard
  :ensure t
  :config
  (dashboard-setup-startup-hook))

(display-line-numbers-mode 1)
(setq display-line-numbers-type 'relative)

(use-package winum
  :init
  (setq winum-keymap
	(let ((map (make-sparse-keymap)))
	  (define-key map (kbd "C-`") 'winum-select-window-by-number)
	  (define-key map (kbd "C-Â²") 'winum-select-window-by-number)
	  (define-key map (kbd "M-0") 'winum-select-window-0-or-10)
	  (define-key map (kbd "M-1") 'winum-select-window-1)
	  (define-key map (kbd "M-2") 'winum-select-window-2)
	  (define-key map (kbd "M-3") 'winum-select-window-3)
	  (define-key map (kbd "M-4") 'winum-select-window-4)
	  (define-key map (kbd "M-5") 'winum-select-window-5)
	  (define-key map (kbd "M-6") 'winum-select-window-6)
	  (define-key map (kbd "M-7") 'winum-select-window-7)
	  (define-key map (kbd "M-8") 'winum-select-window-8)
	  map))
  :config
  (winum-mode 1))

(set-face-font 'default "Roboto Mono Light 10")

;; Line cursor and no blink
(set-default 'cursor-type  '(bar . 1))
(blink-cursor-mode 0)

;; No sound
(setq visible-bell t)
(setq ring-bell-function 'ignore)

;; Paren mode is part of the theme
(show-paren-mode t)

;; No fringe but nice glyphs for truncated and wrapped lines
(fringe-mode '(0 . 0))
(defface fallback '((t :family "Fira Code Light"
                       :inherit 'face-faded)) "Fallback")

(use-package doom-themes
  :config
  ;; Global settings (defaults)
  (setq doom-themes-enable-bold t    ; if nil, bold is universally disabled
        doom-themes-enable-italic t) ; if nil, italics is universally disabled
  (load-theme 'doom-one t)
  ;; or for tremacs users
  (setq doom-themes-treemacs-theme "doom-colors") ; use the colorful treemacs theme
  (doom-themes-treemacs-config)
  ;; Corrects (and improves) org-mode's native fontification.
  (doom-themes-org-config))
